aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp

```diff
@@ -204,17 +209,186 @@ class QLinearDynamicInt8 final : public torch::OperatorKernel {
 
     return output;
   }
-#else // USE_FBGEMM
-  at::Tensor operator()(
-      at::Tensor /* input */,
-      at::Tensor /* packed_weight */) {
-    // We make a strong guarantee that models using these operators will have
-    // the same numerics across different machines. Therefore, we do not provide
-    // a fallback path and rather fail loudly if we cannot run FBGEMM.
+#endif // USE_FBGEMM
+#ifdef USE_PYTORCH_QNNPACK
+
+  struct qparams {
+    float scale;
+    std::int32_t zero_point;
+  };
+
+  qparams chooseQuantizationParams(
+      float min,
+      float max,
+      int32_t qmin,
+      int32_t qmax) {
+    double scale =
+        (std::max(max, 0.f) - std::min(min, 0.f)) / ((double)qmax - qmin);
+    if (scale == 0) {
+      scale = 0.1;
+    }
+    // If scale is 0, we arbitrary adjust the scale to 0.1
+    assert(scale > 0);
+    // We extend the [min, max] interval to ensure that it contains 0.
+    // Otherwise, we would not meet the requirement that 0 be an exactly
+    // representable value.
+    min = std::min(min, 0.f);
+    max = std::max(max, 0.f);
+
+    double zero_point_from_min = qmin - min / scale;
+    double zero_point_from_max = qmax - max / scale;
+    double zero_point_from_min_error = std::abs(qmin) + std::abs(min / scale);
+    double zero_point_from_max_error = std::abs(qmax) + std::abs(max / scale);
+    double initial_zero_point =
+        zero_point_from_min_error < zero_point_from_max_error
+        ? zero_point_from_min
+        : zero_point_from_max;
+
+    int32_t nudged_zero_point = 0;
+    if (initial_zero_point < qmin) {
+      nudged_zero_point = qmin;
+    } else if (initial_zero_point > qmax) {
+      nudged_zero_point = qmax;
+    } else {
+      nudged_zero_point = nearbyint(initial_zero_point);
+    }
```

Comment:
Can you give an example when this part is necessary? I cannot see the necessity of nudging the ZP if we picked the scale properly.

Reply:
It is possible for the initial_zero_point computed to be non-integral. let's say min is 5.5 and max is 555.5 - In this case the computed zero point will have to be nudged to get an integral value.

Reply:
Not really, the nudging cases will only be necessary if we don't have guarantees of zero being representable, but we do (kinda). Here is my logic:

Suppose `initial_zero_point = qmin - min / scale`. We guarantee that `min` is either negative or zero, which makes `initial_zero_point < qmin` to be always false. If we suppose that `initial_zero_point = qmax - max / scale`, the logic is the same: `max` is either positive or zero. Hence, `initial_zero_point > qmax` is always false, so the second condition in the if statement will never be executed.

We can discuss it offline :)